cpu/stm32/periph/ptp.c

```diff
@@ -0,0 +1,216 @@
+     * value in ETH->PTPSSIR to be added onto the subsection register on 
+     * overflow) */
+    ETH->PTPTSAR = adjusted_ptptsar;
+    /* Wait for pending clock speed adjustments to complete */
+    while (ETH->PTPTSCR & ETH_PTPTSCR_TSARU) { }
+    /* Load new PTPTSAR value to hardware */
+    ETH->PTPTSCR |= ETH_PTPTSCR_TSARU;
+    DEBUG("[periph_ptp] Using PTPSAR = %" PRIu32 ", PTPSSIR = %" PRIu32 "\n",
+          adjusted_ptptsar, ptpssir);
+}
+
+void ptp_clock_adjust(int64_t offset)
+{
+    unsigned state = irq_disable();
+    ptp_timestamp_t ts;
+    uint64_t abs_offset = (offset < 0) ? -offset : offset;
+    ptp_ns2ts(&ts, abs_offset);
+    ETH->PTPTSHUR = ts.seconds;
+    ETH->PTPTSLUR = (offset < 0) ? (1UL << 31) | ts.nanoseconds : ts.nanoseconds;
+    while (ETH->PTPTSCR & (ETH_PTPTSCR_TSSTU | ETH_PTPTSCR_TSSTI)) {
+        /* wait until new time value can be set */
+    }
+    ETH->PTPTSCR |= ETH_PTPTSCR_TSSTU;
+    irq_restore(state);
+    DEBUG("[periph_ptp] Updated time by %c%" PRIu32 ".%09" PRIu32 "\n",
+          (offset < 0) ? '-' : '+', (uint32_t)ts.seconds, ts.nanoseconds);
+}
+
+void ptp_clock_set(const ptp_timestamp_t *time)
+{
+    assert(time && time->nanoseconds < NS_PER_SEC);
+    unsigned state = irq_disable();
+    /* First, set the timestamp update registers */
+    ETH->PTPTSHUR = time->seconds;
+    ETH->PTPTSLUR = time->nanoseconds;
+    /* From the data sheet (regarding setting TSSTI):
+     * > Both the TSSTU and TSSTI bits must be read as zero before you can set
+     * > this bit.
+     */
+    while (ETH->PTPTSCR & (ETH_PTPTSCR_TSSTU | ETH_PTPTSCR_TSSTI)) {
+        /* wait until new time value can be set */
+    }
+    /* Now, ask the peripheral to atomically set the clock from the update
+     * registers */
+    ETH->PTPTSCR |= ETH_PTPTSCR_TSSTI;
+    irq_restore(state);
+}
+
+void ptp_clock_read(ptp_timestamp_t *timestamp)
+{
+    /* Read first high register, then low, then again high. If the value in
+     * high register changed between the reads, we start again to prevent
+     * corrupted timestamps being passed to the user. The high register updates
+     * only once per second. So unless the calling thread is interrupted and
+     * blocked for >= 1 second, this should succeed at least on the second try.
+     */
+    do {
+        timestamp->seconds = ETH->PTPTSHR;
+        timestamp->nanoseconds = ETH->PTPTSLR;
+    } while (timestamp->seconds != ETH->PTPTSHR);
+
+    /* TODO: Most significant bit of ETH->PTPTSLR is the sign bit of the time
+     * stamp. Because the seconds register is unsigned, an overflow is not
+     * expected before year 2106. It is not clear from the data sheet, how the
+     * time stamp is to be interpreted when the negative bit is set. For now,
+     * we just ignore this potential source of problems. */
+}
+
+#if IS_USED(MODULE_PERIPH_PTP_TIMER)
+void ptp_timer_clear(void)
+{
+    /* TODO: Use atomic utils, so that bit-banding is used where supported */
```

Comment:
`bit_clear32()` from `bit.h` should also work.

Reply:
Done. I should still switch to atomic utils once available, to have the handling of bit-banding being supported abstracted away.